src/renderers/webgl/WebGLMorphtargets.js

```diff
@@ -27,115 +34,209 @@ function WebGLMorphtargets( gl ) {
 
 		const objectInfluences = object.morphTargetInfluences;
 
-		// When object doesn't have morph target influences defined, we treat it as a 0-length array
-		// This is important to make sure we set up morphTargetBaseInfluence / morphTargetInfluences
+		if ( capabilities.isWebGL2 ) {
 
-		const length = objectInfluences === undefined ? 0 : objectInfluences.length;
+			// instead of using attributes, the WebGL 2 code path encodes morph targets
+			// into an array of data textures. Each layer represents a single morph target.
+			// Note: This texture is created once per geoemtry. If the set of morph targets
+			// changes over time (e.g. by adding additional morph targets), it's necessary
+			// to call dispose() on the geometry and then create a new one.
 
-		let influences = influencesList[ geometry.id ];
+			let texture = morphTextures.get( geometry );
 
-		if ( influences === undefined || influences.length !== length ) {
+			if ( texture === undefined ) {
 
-			// initialise list
+				const hasMorphNormals = geometry.morphAttributes.normal !== undefined;
 
-			influences = [];
+				const morphTargets = geometry.morphAttributes.position;
+				const morphNormals = geometry.morphAttributes.normal || [];
 
-			for ( let i = 0; i < length; i ++ ) {
+				const numberOfMorphTargets = morphTargets.length;
+				const numberOfVertices = geometry.attributes.position.count;
 
-				influences[ i ] = [ i, 0 ];
+				const numberOfVertexData = ( hasMorphNormals === true ) ? 2 : 1; // (v,n) vs. (v)
 
-			}
+				let size = Math.sqrt( numberOfVertices * numberOfVertexData );
+				size = MathUtils.ceilPowerOfTwo( size );
```

Comment:
would not ceilPowerOfTwo waste memory?

Reply:
Do you mind explaining in more detail? I'm using the same approach for computing the size of the bone texture in context of skinning.

https://github.com/mrdoob/three.js/blob/0179baddfcb692931256e6d8cea0f7c4c48a2a7e/src/objects/Skeleton.js#L174

Reply:
for example if the size came out to be 2049 the ceilPowerOfTwo would give the size = 4096 and then the buffer size calculated below would be : 12884.901888 Megabyte ( 4096*4096*3*64 bits) // assumed 64 morph targets

And if it remains 2049 then it would be : 3224.371968 Megabyte ( 4x size reduction in buffer)

Reply:
So ceil should be enough I guess? or did I miss something. This is assuming we are using webgl2 and not webgl1 as webgl2 does not support texture sizes that are not of power 2.

Reply:
Using `ceil()` breaks the logic since the vertex displacement in `webgl_animation_skinning_morph` is now wrong. I'll have a look at this tomorrow since it seems the shader code needs an update.

Reply:
> as webgl1 does not support texture sizes that are not of power 2.

Actually WebGL 1 does support NPOT textures. However, mipmapping does not work with it.

Reply:
Strange. When I compute the texture size like so:
```js
const size = Math.ceil( Math.sqrt( numberOfVertices * numberOfVertexData ) );
```
it produces a rendering issue in `webgl_animation_skinning_morph` that looks like so:

![image](https://user-images.githubusercontent.com/12612165/128673724-67d02732-7650-41ce-a06d-aae28c626725.png)

However, this only happens on macOS. The same code works without issues on Windows 10. I wonder if this is some kind of driver issue?

I guess it would be helpful if somebody else with a macOS could verify the rendering issue.

Reply:
>When I compute the texture size like so:
> const size = Math.ceil( Math.sqrt( numberOfVertices * numberOfVertexData ) );

Works fine for me on macOS 11.5, Chrome, Firefox, Safari. (Only tested default GUI settings.)

Reply:
>  (Only tested default GUI settings.)

Default settings work for me, too. However setting `Angry` morph target to `1` breaks it for me.

Reply:
Still no artifacts for me... It may be helpful if you create a build for easy testing by others.

Reply:
Good idea. Below a link to the example using the optimized texture size computation.

https://raw.githack.com/Mugen87/three.js/dev3/examples/webgl_animation_skinning_morph.html

Reply:
No artifacts on a Pixel 4a and Chrome 🤔 .

Reply:
No artifacts for me on 2015 macbook air running on Mac OS 11.4

Reply:
Thanks for testing! I could imagine that this issue is specific to my iMac GPU (which is from 2014^^) and the respective driver. However, it would be good to get more feedback in order to find out if others experience the issue, too.

It might be worth to file a bug at https://bugs.chromium.org/p/chromium/issues/list. However, the above example is too complex for a test case so it would be necessary to isolate the issue in a smaller demo.